frontend/javascripts/dashboard/dataset/queries.tsx

```diff
@@ -0,0 +1,435 @@
+      onMutate: ([folderId, newParentId]) => {
+        // Optimistically update the folder with the new parent.
+        const previousFolders = queryClient.getQueryData(mutationKey);
+        queryClient.setQueryData(mutationKey, (oldItems: FlatFolderTreeItem[] | undefined) =>
+          (oldItems || []).map((oldFolder: FlatFolderTreeItem) =>
+            oldFolder.id === folderId
+              ? {
+                  ...oldFolder,
+                  parent: newParentId,
+                }
+              : oldFolder,
+          ),
+        );
+        return {
+          previousFolders,
+        };
+      },
+      onSuccess: (updatedFolder, [folderId, newParentId]) => {
+        // Use the returned updatedFolder to update the query data
+        queryClient.setQueryData(mutationKey, (oldItems: FlatFolderTreeItem[] | undefined) =>
+          (oldItems || []).map((oldFolder: FlatFolderTreeItem) =>
+            oldFolder.id === updatedFolder.id
+              ? {
+                  ...updatedFolder,
+                  parent: newParentId,
+                }
+              : oldFolder,
+          ),
+        );
+        queryClient.setQueryData(["folders", folderId], () => updatedFolder);
+      },
+      onError: (err: any, _params, context) => {
+        // Restore the old folder tree. Note that without the optimistic update
+        // and the data reversion here, the sidebar would still show the (incorrectly)
+        // moved folder, because the <SortableTree /> component does its own kind of
+        // optimistic mutation by updating its state immediately when dragging a folder.
+        handleGenericError(err, "Could not update folder. Check the console for details");
+
+        if (context) {
+          queryClient.setQueryData(mutationKey, context.previousFolders);
+        }
+      },
+    },
+  );
+}
+
+export function useUpdateDatasetMutation(folderId: string | null) {
+  const queryClient = useQueryClient();
+  const mutationKey = ["datasetsByFolder", folderId];
+
+  return useMutation(
+    (params: [APIMaybeUnimportedDataset, string] | APIDatasetId) => {
+      // If a APIDatasetId is provided, simply refetch the dataset
+      // without any mutation so that it gets reloaded effectively.
+      if ("owningOrganization" in params) {
+        const datasetId = params;
+        return getDataset(datasetId);
+      }
+      const [dataset, newFolderId] = params;
+      return updateDataset(dataset, dataset, newFolderId, true);
+    },
+    {
+      mutationKey,
+      onSuccess: (updatedDataset) => {
+        console.log("setQueryData for", mutationKey);
+        queryClient.setQueryData(mutationKey, (oldItems: APIMaybeUnimportedDataset[] | undefined) =>
+          updateDatasetInQueryData(updatedDataset, folderId, oldItems),
+        );
+        const targetFolderId = updatedDataset.folderId;
+        if (targetFolderId !== folderId) {
+          // The dataset was moved to another folder. Add the dataset to that target folder
```

Comment:
Yes! This happens above line 314. The code called `updateDatasetInQueryData` to do that, however, I just inlined the code since the additional function didn't really make sense (originally, I refactored it, because I thought I could reuse the code).